### ISSUE TITLE
nvBWT crashing

### ISSUE THREAD

**Issue by psteinb** [NVlabs/nvbio-gpl#5]
Sorry to be a bug today. I build nvBio commit 3e6efbdf5f8e6ac69beaed2f78bb64ac5aab3e09. I am running on a system with 2 GPUs:

``` bash
$ nvidia-smi
+------------------------------------------------------+
| NVIDIA-SMI 5.325.15   Driver Version: 325.15         |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro NVS 295      Off  | 0000:03:00.0     N/A |                  N/A |
| N/A   69C  N/A     N/A /  N/A |       39MB /   255MB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K20c          Off  | 0000:04:00.0     Off |                    0 |
| 33%   45C    P8    27W / 225W |       13MB /  4799MB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Compute processes:                                               GPU Memory |
|  GPU       PID  Process name                                     Usage      |
|=============================================================================|
|    0            Not Supported                                               |
+-----------------------------------------------------------------------------+
```

using 
centos 6.4
cuda 5.5.22
cmake 2.8.11.

Calling

``` bash
nvBWT -d 0 ./RR891269_GM12878_50000cells_1_top16K.fasta ./RR891269_GM12878_50000cells_1_top16K.index
```

results in:

```
info    : buffering bps... started
info    :   buffering "./RR891269_GM12878_50000cells_1_top16K.fasta"
info    : buffering bps... done
info    : 
info    : building forward BWT... started
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  unspecified launch failure
```

If need be, I can provide the input file upon request.
Best -

**Comment by achacond** [comment]
Hello Peter,

I had the same problem than you in a cluster system with a resource manager (slurm).
I show to you what finally worked for me in the 0.9.5:
1) Reserve all the node in the cluster resource manager (all GPUs) and
don't specify in which GPU I want to run the nvBWT (the device flag -d 0).
2) Compile nvbio in release mode (without -g, -lineinfo or -O0 flags).
3) Remove all the scores "-" and underscores "_" in the input/output
filename parameters.

I hope this will be useful for you.

Alejandro Chacón

2014-04-23 15:19 GMT+02:00 Peter Steinbach notifications@github.com:

**Comment by psteinb** [comment]
Hey Alejandro,

thanks for the prompt reply. first the good news, your suggestions of dropping the "-d" flag and replacing all dashes by underscores seems to work perfectly. Is this documented somewhere?

one question to you, from where did you get version 0.9.5?
peter

**Comment by achacond** [comment]
Hello Peter, 

I'm glad that my suggestions worked also for you. 
It's not documented. I have been playing with the library nvbio since 1/2 weeks ago, and I did it work reading the code and by trial and error : ). I was able to index a whole human genome.

I got the 0.9.5 in the repository from here:
https://github.com/NVlabs/nvbio/archive/master.zip

I think that this is the 0.9.5 version, if I'm wrong please make me know.

Alejandro Chacón

**Comment by nsubtil** [comment]
Running debug builds of nvbio is not really recommended, since they're extremely slow. This discussion makes me think that we need to modify the build system to build release by default.

I'm a little concerned about the reports on underscores and similar characters in the file names causing trouble. I've been unable to reproduce this locally, and looking through the argument parsing code it doesn't look like we're doing anything that would handle "special" characters badly, except perhaps for the case where a file name starts with '-' (which I don't believe is the case here).

Can either of you confirm if the file renames are really necessary here?

Thanks!

**Comment by psteinb** [comment]
Revisiting my setup, I ran with a .fasta file including '-' and one with "_" using commit 3e6efbdf5. The error appears only when using the "-d" flag and seems not to depend on the content of "-". But the issue with the "-d" flag is not so nice.

**Comment by nsubtil** [comment]
psteinb, if you're running with '-d 0' on the same system where you got the nvidia-smi output you pasted earlier, then it sounds like you're picking the Quadro NVS 295 to run on. Given the amount of memory on that GPU, I wouldn't expect nvBWT to run successfully.

It should run successfully with '-d 1'. If you drop the '-d' entirely, then it should default to running on the K20c and also run successfully.

**Comment by psteinb** [comment]
hello, I checked again:

``` bash
$ nvBWT -d 0 RR891269_GM12878_50000cells_1_top16K.fasta RR891269_GM12878_50000cells_1_top16K-index
info    : max length : 0
info    : input      : "RR891269_GM12878_50000cells_1_top16K.fasta"
info    : output     : "RR891269_GM12878_50000cells_1_top16K-index"
verbose :   cuda devices : 2
verbose :   chosen device 0
verbose :     device name        : Tesla K20c
verbose :     compute capability : 3.5
...
info    : building forward BWT... started
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  unspecified launch failure
Aborted
```

and 

``` bash
$ nvBWT -d 1 RR891269_GM12878_50000cells_1_top16K.fasta RR891269_GM12878_50000cells_1_top16K-index
info    : max length : 1
info    : input      : "RR891269_GM12878_50000cells_1_top16K.fasta"
info    : output     : "RR891269_GM12878_50000cells_1_top16K-index"
verbose :   cuda devices : 2
verbose :   chosen device 1
verbose :     device name        : Quadro NVS 295
verbose :     compute capability : 1.1
....
info    : buffering bps... done
error   : caught a std::runtime_error exception:
error   : 
```

As reported earlier, omitting the "-d" flag let's the tool run through alright.
Here is the input file for reference: http://idisk.mpi-cbg.de/~steinbac/RR891269_GM12878_50000cells_1_top16K.fasta

**Comment by jpantaleoni** [comment]
Judging from your log it looks like the -d option is wrongly taken to limit
the length of the input string (max-length: 0) - we'll take a look.

**Comment by nsubtil** [comment]
Fixed with commit 21e8c21a80
